src/context.rs

```diff
@@ -0,0 +1,213 @@
+use futures::future::{pending, BoxFuture, FutureExt, Shared};
+use tokio::time::Duration;
+
+pub struct CancelHandle {
+    cancel_fn: Box<dyn FnOnce()>,
+}
+
+impl CancelHandle {
```

Comment:
I'm considering getting rid of this and not wrap `AbortHandle` like I'm doing on the `with_cancel` call. I don't see the need for that type now